### PR TITLE
Clear and scroll console buttons

### PR DESCRIPTION
**Pull request by jdeblander**
I implemented 2 extra functionalities for the console in utiLITI, as mentioned in #358.
One to clear the current console and start with a blank slate,
and one to auto-scroll to the bottom of the logs in the current console.
With two small buttons on the left side of the console panel, you can easily perform these actions.

image

Though the functionality is not part of the core usage of the utiLITI tool,
it could be a handy addition for some users when using the built in console.

_One remark, the current icons used on the new buttons could use an extra touch.
It seems I'm not able to create a proper icon that is crisp enough._

### CODE REVIEW

**Comment by jdeblander** [comment]
About the failed test.
For some reason it fails a time and passes another time.
As it includes a JTextPane and other UI components, without a real UI set up, it is possible that the reason lays in here.
I'm looking into a way to make this test more stable.

Any suggestions are welcome.

**Comment by Gamebuster19901** [comment]
I'm unable to get that test to fail on my machine, or on github (https://travis-ci.com/github/Gamebuster19901/litiengine/builds/207842490). 

I have created a PR on your fork to print the stacktrace whenever the test does fail though.

**Comment by jdeblander** [comment]
Well, I merged your PR and the next build is successful.
I'm not really convinced about the stability yet.

**Comment by Gamebuster19901** [comment]
Yeah that's weird. None of the logic has changed, a NPE should still be thrown as none of the logic was changed.

**Review by nightm4re94 — APPROVED** [review]
Thank you very much!
At some point, we might consider making the buttons togglable through a configuration parameter.

File: utiliti/src/de/gurkenlabs/utiliti/swing/LogHandler.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    @Override
+    public void close() {
+        // nothing to close here -> writing to a control
+    }
+
+    public void scrollToLast() {
+        StyledDocument doc = textPane.getStyledDocument();
+        Rectangle2D bounds = null;
+        try {
+            bounds = textPane.modelToView2D(textPane.getCaretPosition());
```
Comment (nightm4re94) [inline utiliti/src/de/gurkenlabs/utiliti/swing/LogHandler.java:80]: I just noticed that [JTextComponent.modelToView2D(int pos)](https://docs.oracle.com/javase/9/docs/api/javax/swing/text/JTextComponent.html#modelToView2D-int-) is incompatible with Java 8. Please keep in mind that LITIENGINE currently builds with Java 8 and cannot utilize any features introduced later on. In this case, we need to use [JTextComponent.modelToView(int pos)](https://docs.oracle.com/javase/8/docs/api/javax/swing/text/JTextComponent.html#modelToView-int-).

**Review by nightm4re94 — COMMENTED** [review]
I've noticed a Java 8 incompatible method use here (as described in the comment).

File: utiliti/src/de/gurkenlabs/utiliti/swing/LogHandler.java (hunk 1 above)
Comment (nightm4re94) [reply utiliti/src/de/gurkenlabs/utiliti/swing/LogHandler.java:80]: This has been fixed by e4a64f8f81686ae5bb189bd6dec57ebee6497a77.

Comment (Gamebuster19901) [reply utiliti/src/de/gurkenlabs/utiliti/swing/LogHandler.java:80]: I've created a PR which makes the Travis build fail if the code is not java 8 compatible.

**Comment by nightm4re94** [comment]
Now that the test failure randomly appeared in the master branch, I will revert the addition of the `LogHandlerTest` class.
Please scan the stacktrace for the error source and file a pull request once the problem has been solved properly.
```java
Exception in thread "AWT-EventQueue-0" java.lang.ArrayIndexOutOfBoundsException
	at java.lang.System.arraycopy(Native Method)
	at javax.swing.text.BoxView.updateLayoutArray(BoxView.java:214)
	at javax.swing.text.BoxView.replace(BoxView.java:185)
	at javax.swing.text.View.append(View.java:417)
	at javax.swing.text.FlowView$FlowStrategy.layout(FlowView.java:473)
	at javax.swing.text.FlowView.layout(FlowView.java:201)
	at javax.swing.text.BoxView.setSize(BoxView.java:397)
	at javax.swing.text.BoxView.updateChildSizes(BoxView.java:366)
	at javax.swing.text.BoxView.setSpanOnAxis(BoxView.java:348)
	at javax.swing.text.BoxView.layout(BoxView.java:708)
	at javax.swing.text.BoxView.setSize(BoxView.java:397)
	at javax.swing.plaf.basic.BasicTextUI$RootView.setSize(BasicTextUI.java:1722)
	at javax.swing.plaf.basic.BasicTextUI.modelToView(BasicTextUI.java:1046)
	at javax.swing.text.DefaultCaret.repaintNewCaret(DefaultCaret.java:1308)
	at javax.swing.text.DefaultCaret$1.run(DefaultCaret.java:1287)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at javax.swing.text.BoxView.calculateMajorAxisRequirements(BoxView.java:871)
	at javax.swing.text.BoxView.checkRequests(BoxView.java:930)
	at javax.swing.text.BoxView.getMinimumSpan(BoxView.java:568)
	at javax.swing.text.BoxView.calculateMajorAxisRequirements(BoxView.java:871)
	at javax.swing.text.BoxView.checkRequests(BoxView.java:930)
	at javax.swing.text.BoxView.setSpanOnAxis(BoxView.java:329)
	at javax.swing.text.BoxView.layout(BoxView.java:709)
	at javax.swing.text.BoxView.setSize(BoxView.java:397)
	at javax.swing.plaf.basic.BasicTextUI$RootView.setSize(BasicTextUI.java:1722)
	at javax.swing.plaf.basic.BasicTextUI.modelToView(BasicTextUI.java:1046)
	at javax.swing.text.DefaultCaret.repaintNewCaret(DefaultCaret.java:1308)
	at javax.swing.text.DefaultCaret$1.run(DefaultCaret.java:1287)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)

java.lang.ArrayIndexOutOfBoundsException
	at java.lang.System.arraycopy(Native Method)
	at javax.swing.text.BoxView.updateLayoutArray(BoxView.java:214)
	at javax.swing.text.BoxView.replace(BoxView.java:185)
	at javax.swing.text.View.append(View.java:417)
	at javax.swing.text.FlowView$FlowStrategy.layout(FlowView.java:473)
	at javax.swing.text.FlowView.layout(FlowView.java:201)
	at javax.swing.text.BoxView.setSize(BoxView.java:397)
	at javax.swing.text.BoxView.updateChildSizes(BoxView.java:366)
	at javax.swing.text.BoxView.setSpanOnAxis(BoxView.java:348)
	at javax.swing.text.BoxView.layout(BoxView.java:708)
	at javax.swing.text.BoxView.setSize(BoxView.java:397)
	at javax.swing.plaf.basic.BasicTextUI$RootView.setSize(BasicTextUI.java:1722)
	at javax.swing.plaf.basic.BasicTextUI.modelToView(BasicTextUI.java:1046)
	at javax.swing.plaf.basic.BasicTextUI.modelToView(BasicTextUI.java:1022)
	at javax.swing.text.JTextComponent.modelToView(JTextComponent.java:1377)
	at de.gurkenlabs.utiliti.swing.LogHandler.scrollToLast(LogHandler.java:80)
	at de.gurkenlabs.utiliti.swing.LogHandlerTest.scrollToLast(LogHandlerTest.java:66)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:628)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:117)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:184)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:180)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:127)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:132)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:412)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)

LogHandlerTest > scrollToLast() FAILED
```

**Comment by Gamebuster19901** [comment]
Looks like a possible threading issue. Is the test being run on a AWT/Swing thread?

http://opensha.usc.edu/trac/ticket/379

**Pull request merged**